sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/io/aws2/kinesis/enhancedfanout/EFOShardSubscribersPool.java

```diff
@@ -196,33 +200,45 @@ public void start(Iterable<ShardCheckpoint> checkpoints) {
    */
   @Nullable
   KinesisRecord getNextRecord() throws IOException {
-    if (current == null && eventQueue.isEmpty()) {
-      if (subscriptionError == null) {
-        return null;
-      } else {
-        throw new IOException(subscriptionError);
+    while (true) {
+      if (!isStopped && subscriptionError != null) {
+        // Stop the pool to cancel all subscribers and prevent new subscriptions.
+        // Doing this as part of getNextRecord() avoids concurrent access to the state map and
+        // prevents any related issues.
+        stop();
       }
-    }
 
-    if (current == null) {
-      current = eventQueue.poll();
-    }
+      if (current == null) {
+        current = eventQueue.poll();
+      }
 
-    if (current != null) {
-      String shardId = current.shardId;
-      ShardState shardState = Preconditions.checkStateNotNull(state.get(shardId));
-      if (current.hasNext()) {
-        KinesisClientRecord r = current.next();
-        shardState.update(r);
-        KinesisRecord kinesisRecord = new KinesisRecord(r, read.getStreamName(), shardId);
-        latestRecordTimestampPolicy.update(kinesisRecord);
-        return kinesisRecord;
+      if (current != null) {
+        String shardId = current.shardId;
+        ShardState shardState = Preconditions.checkStateNotNull(state.get(shardId));
+        if (current.hasNext()) {
+          KinesisClientRecord r = current.next();
+          shardState.update(r);
+          // Make sure to update shard state accordingly if `current` does not contain any more
```

Comment:
Looks like you've removed this. Was there any problem with it?
As far as I understand this is important to make sure any checkpoint taken at this point reflects the new shard state, e.g. of the current shard is closed and replaced by successors.

Reply:
Removing it did not bring any problems, lost records or dups with re-sharding. My implementation of `getNextRecord()` assumed that we have a timeline like this:
- call 1 of `getNextRecord()` - we return last record of `current`
- call 2 of `getNextRecord()` - we come to line 232, update state and then return `null`